File: backend/src/app/modules/accounts/services/profile_service.py
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.modules.accounts.models.user import User
from app.modules.accounts.models.user_profile import UserProfileRolePreference, UserProfileShipPreference
from app.modules.accounts.schemas.profile_update import ProfileUpdate
from app.modules.fleet.services.fleet_service import sync_user_primary_fleet
from app.modules.permissions.models.role import FleetRoleDefinition
from app.modules.ships.models.ship import Ship


class ProfileValidationError(ValueError):
    pass


def _validate_ids(db: Session, payload: ProfileUpdate) -> None:
    ship_ids = set(db.scalars(select(Ship.id).where(Ship.id.in_(payload.preferred_ship_ids), Ship.is_active.is_(True))).all())
    role_ids = set(db.scalars(select(FleetRoleDefinition.id).where(FleetRoleDefinition.id.in_(payload.preferred_role_ids))).all())
    if ship_ids != set(payload.preferred_ship_ids):
        raise ProfileValidationError("One or more preferred ships are invalid.")
    if role_ids != set(payload.preferred_role_ids):
        raise ProfileValidationError("One or more preferred roles are invalid.")


def update_profile(db: Session, user: User, payload: ProfileUpdate) -> User:
    """Update user-owned profile data and normalized preferences."""

    _validate_ids(db, payload)
    user.display_name = payload.display_name
    sync_user_primary_fleet(db, user)
    if user.fleet_id is None:
        user.fleet_name = payload.fleet_name
    profile = user._ensure_profile()
    profile.preferred_focus = payload.preferred_focus
    profile.availability = payload.availability
    profile.timezone = payload.timezone
    profile.discord_handle = payload.discord_handle
    profile.note = payload.note
    profile.ship_preferences = [
        UserProfileShipPreference(ship_id=ship_id, sort_order=index * 10)
        for index, ship_id in enumerate(payload.preferred_ship_ids, start=1)
    ]
    profile.role_preferences = [
        UserProfileRolePreference(fleet_role_id=role_id, sort_order=index * 10)
        for index, role_id in enumerate(payload.preferred_role_ids, start=1)
    ]
    db.add(user)
    db.commit()
    db.refresh(user)
    return user

File: spring-api/src/main/java/eu/royalblackwater/api/account/ProfileService.java
package eu.royalblackwater.api.account;

import eu.royalblackwater.api.contract.ProfileUpdate;
import eu.royalblackwater.api.fleet.FleetMembershipRepository;
import eu.royalblackwater.api.persistence.JdbcQueryService;
import java.time.Clock;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import static org.springframework.http.HttpStatus.UNPROCESSABLE_ENTITY;

@Service
public class ProfileService {
    private static final Set<String> FOCUS_VALUES = Set.of(
            "pve_farming", "pve_imp_hunting", "pve_general", "pvp_open_world",
            "pvp_arena", "pvp_general", "trading", "other");

    private final UserRepository users;
    private final FleetMembershipRepository memberships;
    private final JdbcQueryService jdbc;
    private final UserViewService views;
    private final Clock clock;

    public ProfileService(UserRepository users, FleetMembershipRepository memberships, JdbcQueryService jdbc,
                          UserViewService views, Clock clock) {
        this.users = users;
        this.memberships = memberships;
        this.jdbc = jdbc;
        this.views = views;
        this.clock = clock;
    }

    @Transactional
    public AuthContracts.UserRead update(int userId, ProfileUpdate payload) {
        UserEntity user = users.findById(userId)
                .orElseThrow(() -> new IllegalStateException("Authenticated user no longer exists."));
        List<Integer> shipIds = normalizedIds(payload.preferredShipIds(), 20);
        List<Integer> roleIds = normalizedIds(payload.preferredRoleIds(), 10);
        validateIds(shipIds, roleIds);
        String focus = normalized(payload.preferredFocus());
        if (focus != null && !FOCUS_VALUES.contains(focus)) {
            throw invalid("Invalid preferred focus.");
        }
        LocalDateTime now = LocalDateTime.ofInstant(clock.instant(), ZoneOffset.UTC);
        UserProfileEntity profile = user.ensureProfile(now);
        String externalFleet = memberships.existsByUser_Id(userId) ? profile.getExternalFleetName() : normalized(payload.fleetName());
        profile.update(payload.displayName().strip(), externalFleet, focus, normalized(payload.availability()),
                normalized(payload.timezone()), normalized(payload.discordHandle()), normalized(payload.note()), now);
        profile.replaceShipPreferences(shipIds);
        profile.replaceRolePreferences(roleIds);
        user.touch(now);
        users.save(user);
        return views.read(userId);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> options() {
        return Map.of(
                "ships", jdbc.query("""
                        select id, name, rate
                          from ships
                         where is_active = true
                         order by rate, name, id
                        """, Map.of()),
                "roles", jdbc.query("""
                        select id, code, label
                          from fleet_roles
                         order by rank desc, label, id
                        """, Map.of()));
    }

    private void validateIds(List<Integer> shipIds, List<Integer> roleIds) {
        if (!shipIds.isEmpty()) {
            long count = jdbc.count("select count(*) from ships where is_active = true and id in (:ids)", Map.of("ids", shipIds));
            if (count != shipIds.size()) throw invalid("One or more preferred ships are invalid.");
        }
        if (!roleIds.isEmpty()) {
            long count = jdbc.count("select count(*) from fleet_roles where id in (:ids)", Map.of("ids", roleIds));
            if (count != roleIds.size()) throw invalid("One or more preferred roles are invalid.");
        }
    }

    private static List<Integer> normalizedIds(List<Long> values, int maximum) {
        if (values == null || values.isEmpty()) return List.of();
        LinkedHashSet<Integer> ids = new LinkedHashSet<>();
        for (Long value : values) {
            if (value == null || value < 1 || value > Integer.MAX_VALUE) throw invalid("Invalid preference identifier.");
            ids.add(value.intValue());
        }
        if (ids.size() > maximum) throw invalid("Too many preferences.");
        return List.copyOf(ids);
    }

    private static String normalized(String value) {
        if (value == null) return null;
        String stripped = value.strip();
        return stripped.isEmpty() ? null : stripped;
    }

    private static ResponseStatusException invalid(String detail) {
        return new ResponseStatusException(UNPROCESSABLE_ENTITY, detail);
    }
}
